src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +112,494 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+                        return;
+                    }
+
+                    var disposeInterfaceMethod = disposableType.GetMembers(DisposeMethodName).Single() as IMethodSymbol;
+                    if (disposeInterfaceMethod == null)
+                    {
+                        return;
+                    }
+
+                    var garbageCollectorType = context.Compilation.GetTypeByMetadataName("System.GC");
+                    if (garbageCollectorType == null)
+                    {
+                        return;
+                    }
+
+                    var suppressFinalizeMethod = garbageCollectorType.GetMembers("SuppressFinalize").Single() as IMethodSymbol;
+                    if (suppressFinalizeMethod == null)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new Analyzer(context.Compilation, disposableType, disposeInterfaceMethod, suppressFinalizeMethod);
+                    analyzer.Initialize(context);
+                });
+        }
+
+        private static bool IsDisposeBoolMethod(IMethodSymbol method)
+        {
+            if (method.Name == DisposeMethodName && method.MethodKind == MethodKind.Ordinary &&
+                method.ReturnsVoid && method.Parameters.Length == 1)
+            {
+                var parameter = method.Parameters[0];
+                if (parameter.Type != null && parameter.Type.SpecialType == SpecialType.System_Boolean && parameter.RefKind == RefKind.None)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Analyzes single instance of compilation.
+        /// </summary>
+        private class Analyzer
+        {
+            private readonly Compilation compilation;
+            private readonly INamedTypeSymbol disposableType;
+            private readonly IMethodSymbol disposeInterfaceMethod;
+            private readonly IMethodSymbol suppressFinalizeMethod;
+
+            public Analyzer(Compilation compilation, INamedTypeSymbol disposableType, IMethodSymbol disposeInterfaceMethod, IMethodSymbol suppressFinalizeMethod)
+            {
+                this.compilation = compilation;
+                this.disposableType = disposableType;
+                this.disposeInterfaceMethod = disposeInterfaceMethod;
+                this.suppressFinalizeMethod = suppressFinalizeMethod;
+            }
+
+            public void Initialize(CompilationStartAnalysisContext context)
+            {
+                context.RegisterSymbolAction(AnalyzeNamedTypeSymbol, SymbolKind.NamedType);
+                context.RegisterOperationBlockAction(AnalyzeOperationBlock);
+            }
+
+            private void AnalyzeNamedTypeSymbol(SymbolAnalysisContext context)
+            {
+                var type = context.Symbol as INamedTypeSymbol;
+                if (type != null && type.IsType && type.TypeKind == TypeKind.Class)
```

Comment:
since you already have a namedTypeSymbol, the `type.IsType` checks is not needed.